util: move utf8 utils to a separate header

fabiobaltieri opened the pull request:
Hi, bumped into a build regression since https://github.com/zephyrproject-rtos/zephyr/pull/94779, one of our downstream application uses an interesting combination of configs that triggered a circular dependency build failure (which was a lot of fun to troubleshoot).

Managed to reproduce the issue upstream with

```diff
diff --git a/samples/basic/blinky/src/main.c b/samples/basic/blinky/src/main.c
index 4cab4969d94..5ebcbcf8a97 100644
--- a/samples/basic/blinky/src/main.c
+++ b/samples/basic/blinky/src/main.c
@@ -4,6 +4,9 @@
  * SPDX-License-Identifier: Apache-2.0
  */
 
+#define _GNU_SOURCE
+#include <sys/types.h>
+
 #include <stdio.h>
 #include <zephyr/kernel.h>
 #include <zephyr/drivers/gpio.h>
```

and `west build -p -b nrf52dk/nrf52832 samples/basic/blinky -DCONFIG_POSIX_API=y -DCONFIG_NEWLIB_LIBC=y`

May look into adding a test case for this but for now I think the right thing to do is to split these on their own header.

--

This moves the declaration of the utf8 utils defined in lib/utils/utf8.c in their own header. Main reason to do this is that the current setup requried adding an include for sys/types.h in util.h, which can result in a build falure due to a circular header depdenecy when using:

CONFIG_POSIX_API=y
CONFIG_NEWLIB_LIBC=y
_GNU_SOURCE

the loop and error are:
- include/sys/types.h:50: <- this is a NEWLIB one
- include/zephyr/posix/sys/select.h:9:
- include/zephyr/posix/posix_types.h:30:
- include/zephyr/kernel.h:17:
- include/zephyr/kernel_includes.h:25:
- include/zephyr/sys/atomic.h:18:

include/zephyr/sys/util.h:705:1: error: unknown type name 'ssize_t'

fabiobaltieri committed:
util: move utf8 utils to a separate header

This moves the declaration of the utf8 utils defined in lib/utils/utf8.c
in their own header. Main reason to do this is that the current setup
requried adding an include for sys/types.h in util.h, which can result
in a build falure due to a circular header depdenecy when using:

CONFIG_POSIX_API=y
CONFIG_NEWLIB_LIBC=y
_GNU_SOURCE

the loop and error are:

- include/sys/types.h:50: <- this is a NEWLIB one
- include/zephyr/posix/sys/select.h:9:
- include/zephyr/posix/posix_types.h:30:
- include/zephyr/kernel.h:17:
- include/zephyr/kernel_includes.h:25:
- include/zephyr/sys/atomic.h:18:

include/zephyr/sys/util.h:705:1:
error: unknown type name 'ssize_t'

Signed-off-by: Fabio Baltieri <fabiobaltieri@google.com>

fabiobaltieri commented:
cc @iliar

Thalley reviewed (dismissed):
Changes are OK, but does sound like there's an underlying issue that perhaps need to be fixed at some point :)

fabiobaltieri commented:
Flagging as hotfix since there's a chance it broke other users as well and it's not really changing any code per se, just reorganizing the headers.

nashif commented:
I do not think this should be treated as a hotfix, this creates new headers, moves things around etc and has the chance of breaking existing users. Is this blocking anythjing significan anywhere in zephyr upstream?

fabiobaltieri committed:
tests: posix: add a test for types.h and _GNU_SOURCE

This combination (with newlib) seems to catch some build failures due to
include depdenency, adding a test file so it gets caught in CI.

Signed-off-by: Fabio Baltieri <fabiobaltieri@google.com>

fabiobaltieri commented:
CI test for this: https://github.com/zephyrproject-rtos/zephyr/pull/95395

cfriedt reviewed (commented):
@fabiobaltieri - my 2¢: leave the declarations in `sys/util.h` but remove `<sys/types.h>` and change the `ssize_t` to `int`.

The reasoning: Zephyr does occasionally get into trouble using POSIX types, especially the kind that have a length dependent on the arch.

In ISO C, `time_t` and `size_t` fall into that category.

In POSIX, `ssize_t`, `off_t`, and possibly others.

fabiobaltieri commented:
> @fabiobaltieri - my 2¢: leave the declarations in `sys/util.h` but remove `<sys/types.h>` and change the `ssize_t` to `int`.

I can do that, but looking at this file I think that these utf8 utilities should have never been in this file in the first place, util.h should really just be basic utilities that can find use anywhere, these utf8 string analysis/manipulation things belongs to their own header, how about keeping this to move them out and then you can do whatever you want with the type?

Thalley commented on include/zephyr/sys/util_utf8.h:
```diff
@@ -0,0 +1,94 @@
+ * @return Number of UTF-8 characters in @p s on success or (negative) error code
+ *  otherwise.
+ */
+ssize_t utf8_count_chars(const char *s);
```
> I think it's a good idea to have these on their own header regardless, it's a group of function for a specific purpose, chances are it will grow, it has its own C file...

Yeah, not opposed to that :) But it seems worth to still have the discussion on `ssize_t` and other POSIX types in the public header files

iliar replied:
Thanks for catching this. 
Maybe it is a good idea to use just `int`. 
There might be some cases where we do not want to enable POSIX types, but we need UTF functions.

fabiobaltieri replied:
Well those types are available regardless of whether you enable POSIX or not, the problem is that when you enable POSIX they come from a different include file, which ends up including other stuff and triggering the include loop problem I bumped into, but it's not like having `ssize_t` requires you to enable POSIX.

Thalley replied:
Isn't `ssize_t` a POSIX specific type? It's not defined by any C standard AFAIK

From the `man` page
```
STANDARDS
       size_t C11, POSIX.1-2008.

       ssize_t
              POSIX.1-2008.

HISTORY
       size_t C89, POSIX.1-2001.

       ssize_t
              POSIX.1-2001.
```

fabiobaltieri replied:
> we do not want to enable POSIX types

what I'm saying is that it does not appear like you have to enable anything to use it, anyway I'm happy to add a commit to change it if you really don't want to use it anymore, `strlen` uses `size_t`, want me to swap it to that?

Thalley replied:
> strlen uses size_t, want me to swap it to that?

If we want to be able to return errors in case of invalid input (`strlen` doesn't do that), then we need something that can return a negative value. `int` will probably work OK (for any strings up to 2GiB). If we want to support larger strings we could use `int64_t` as the return value, but at this point I'm OK with `int` as well

fabiobaltieri replied:
Alright pushed an extra commit for it.

fabiobaltieri committed:
utils: utf8: change utf8_count_chars return to int

Change utf8_count_chars return type to int and drop thesys/types.h, this
way the function does not depend on posix types.

Signed-off-by: Fabio Baltieri <fabiobaltieri@google.com>

fabiobaltieri commented on include/zephyr/sys/util.h:
```diff
@@ -29,8 +29,6 @@
 #include <stddef.h>
 #include <stdint.h>
 #include <string.h>
-#include <sys/types.h>
-
```
alright dropped

fabiobaltieri commented on include/zephyr/sys/util_utf8.h:
```diff
@@ -0,0 +1,94 @@
+#define ZEPHYR_INCLUDE_SYS_UTIL_UFT8_H_
+
+#include <stddef.h>
+#include <sys/types.h>
```
gone

fabiobaltieri commented:
Ok so this is now a:
- split the header on their own file
- stop using types.h
- add a test for the nasty use case that led to this

@cfriedt happy with me adding that file in the posix tests?

@nashif this is going to fail compliance because `_GNU_SOURCE` so it'll need org superpower to merge

fabiobaltieri commented:
@nashif? (note that there's a CI false positive)
